packages/teleterm/src/ui/utils/retryWithRelogin.ts

```diff
@@ -0,0 +1,71 @@
+import {} from 'teleterm/ui/services/workspacesService/documentsService';
+import { routing } from 'teleterm/ui/uri';
+import AppContext from 'teleterm/ui/appContext';
+
+export async function retryWithRelogin<T>(
+  appContext: AppContext,
+  originatingDocumentUri: string,
+  resourceUri: string,
+  actionToRepeat: () => Promise<T>
+): Promise<T> {
+  let retryableErrorFromActionToRepeat: Error;
+  try {
+    return await actionToRepeat();
+  } catch (error) {
+    // TODO(ravicious): Replace this with actual check on metadata.
+    const isRetryable =
+      error instanceof Error &&
+      (error.message.includes('ssh: handshake failed') ||
+        error.message.includes('ssh: cert has expired'));
+
+    if (isRetryable) {
+      retryableErrorFromActionToRepeat = error;
+    } else {
+      throw error;
+    }
+  }
+
+  const isDocumentStillActive = appContext.workspacesService.isDocumentActive(
+    originatingDocumentUri
+  );
+
+  if (!isDocumentStillActive) {
+    // The error is retryable, but the user is no longer looking at the relevant UI, for example
+    // they switched to a different document or a different workspace completely.
+    //
+    // Since it might take a few seconds before the execution gets to this point, the user might no
+    // longer remember what their intent was, thus showing them a login modal can be disorienting.
+    //
+    // In that situation, let's just not attempt to relogin and instead let's surface the original
+    // error.
+    throw retryableErrorFromActionToRepeat;
+  }
```

Comment:
In case we ever need to use `retryWithRelogin`, this part can be extracted to a separate function that `retryWithRelogin` is going to accept as an arg.

The purpose of the function would be the same as this code here – to check if the user is still seeing the relevant UI.